File: 추천50제/bfs/16236.py
import queue
import sys
from collections import deque

#처음 아기 상어 크기 2
#자신의 크기보다 큰 물고기 칸 못감
#같은 크기는 지나갈순 있지만 먹진못함

#먹을수 있는 고기가 1마리보다 많으면 가장 가까운 물고기
#같은 거리가 여러개라면 가장 위/가장 왼쪽

# 이동하기 전에 먹을수 있는 물고기의 좌표 확인
# 좌표를 y축 기준 정렬 그다음 x축 기준 정렬
# 해당 좌표로 이동
# 물고기를 먹고 레벨업 하면 다시한번 먹을 수 있는 물고기
# 좌표 확인
# 해당 과정 반복

input = sys.stdin.readline
n = int(input())
board = []
for _ in range(n):
    board.append(list(map(int,input().rstrip().split())))

shark_size = 2
time = 0
dx = [0,0,-1,1]
dy = [1,-1,0,0]
cnt_fish = 0
eat_cnt = 0
fish_pos = []
shark_y,shark_x=0,0
for i in range(n):
    for j in range(n):
        if 0<board[i][j]<7:
            cnt_fish+=1
            fish_pos.append([i,j])
        elif board[i][j]==9:
            shark_y,shark_x = i,j
board[shark_y][shark_x]=0

def bfs(shark_x,shark_y):
    queue = deque([[shark_y,shark_x,0]])
    dist_list = []
    visited = [[False]*n for _ in range(n)]
    visited[shark_y][shark_x]=True
    min_dist = int(1e9)
    while queue:
        y,x,dist = queue.popleft()
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if -1<nx<n and -1<ny<n and not visited[ny][nx]:
                if board[ny][nx] <=shark_size:
                    visited[ny][nx]=True
                    if visited[ny][nx]<shark_size:
                        min_dist = dist
                        dist_list.append([dist+1,ny,nx])
                    if dist+1<=min_dist:
                        queue.append([ny,nx,dist+1])
        if dist_list:
            dist_list.sort()
            return dist_list[0]
        else:
            return False
while cnt_fish:
    result = bfs(shark_x,shark_y)
    if not result:
        break
    shark_y,shark_x = result[1],result[2]
    time += result[0]
    eat_cnt+=1
    cnt_fish-=1
    if eat_cnt == shark_size:
        shark_size+=1
        eat_cnt=0
    board[shark_y][shark_x]=0

print(time)

File: Samsung/16236.js
let fs = require("fs");
let input = fs
  .readFileSync("dev/stdin")
  .toString()
  .trim()
  .split("\n")
  .map((e) => e.split(" ").map(Number));
let n = input.shift()[0];
let size = 2;
let exp = 2;
let dx = [-1, 1, 0, 0];
let dy = [0, 0, -1, 1];
let stack = [];
let answer = 0;
//먹을수있는게 없으면 엄마 콜
//먹을수있는 물고기가 여러마리면 가까운곳부터
//가까운 곳이 여러곳이라면 가장 위, 가장 위가 여러마리라면 가장 왼쪽
//가까운 순으로 정렬, 컬럼 순으로 정렬, 로우 순으로 정렬
for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    if (input[i][j] == 9) {
      bfs(i, j);
      input[i][j] = 0;
    }
  }
}
//한번 전체 탐색 하며 정렬
//현재 위치 변경
//반복
//
function bfs(i, j) {
  stack = [];
  let visited = Array.from(Array(n), () => Array(n).fill(0));
  visited[i][j] = 1;
  let q = [[i, j, 0]];
  while (q.length > 0) {
    let [x, y, distance] = q.shift();
    for (let k = 0; k < 4; k++) {
      let nx = x + dx[k];
      let ny = y + dy[k];
      if (
        -1 < nx &&
        nx < n &&
        -1 < ny &&
        ny < n &&
        visited[nx][ny] == 0 &&
        input[nx][ny] <= size
      ) {
        visited[nx][ny] = 1;
        q.push([nx, ny, distance + 1]);
        if (input[nx][ny] != 0 && input[nx][ny] < size) {
          stack.push([nx, ny, distance + 1]);
        }
      }
    }
  }
}

while (stack.length != 0) {
  stack.sort((a, b) => {
    if (a[2] == b[2]) {
      if (a[0] == b[0]) {
        return a[1] - b[1];
      } else {
        return a[0] - b[0];
      }
    } else {
      return a[2] - b[2];
    }
  });
  let [sharkx, sharky, distance] = stack.shift();
  input[sharkx][sharky] = 0; //물고기 먹음
  answer += distance;
  exp -= 1; //경험치++
  if (exp == 0) {
    size++;
    exp = size;
  }
  bfs(sharkx, sharky);
}

console.log(answer);
